### PR TITLE
initial commit to add launchpad and mission control into QMK.

### PR DESCRIPTION
**Pull request by lbibass**
I am ***TRYING*** to add in Mission Control and Launchpad into QMK. 

## Description

I am following what @grota did in his addition of brightness controls in #4477. 

As of right now, it is not working, and I cannot figure out why. 
## Types of changes

- [ ] Core
- [ ] Bugfix
- [x] New Feature
- [ ] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by this PR

* 

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by lbibass** [comment]
Looking at https://www.usb.org/sites/default/files/hutrr77_-_desktop_controls_0.pdf, the keycodes I am using should work. Also, I think that I have done everything that grota did. Not sure as to why it fails to compile.

**Comment by grota** [comment]
In my contribution brightness up and down are two concepts related to the usb hid standard (and the hexadecimal values are also defined by that standard).
Disclaimer, I'm not an expert of Mac stuff, but I'm guessing those are two programs, where are these two values (0x2a0 and 0x29F) coming from? (we might be comparing apples to oranges here)

**Comment by grota** [comment]
I think it's because in the enum KC_MISSION_CONTROL ends up having the value 0xC0 (it's the 16th from KC_MEDIA_EJECT). Not sure how the policy should be here.

**Comment by lbibass** [comment]
yeah, I am just confused at this point. it SHOULD work. But it doesn't. I mean, it works in the official docs!

**Comment by grota** [comment]
No, I mean that just looking at the output of the travis build I think that KC_LAUNCHPAD ends up having the value 0xC0 in the enum like its follower KC_FN0 which might messes things up.
I think the main problem is that there's the assumption that the number of keycodes can only be 255 (0xff) in total. And that's why the values are so contiguous. Someone more informed and expert than me should look at it.

**Comment by fauxpark** [comment]
There is now only one free spot left after the screen brightness keycodes (`0xBF`). The rest of the space has been already filled.

If you're looking for a way to make your keyboard more Apple-y, have a look at https://github.com/qmk/qmk_firmware/issues/2179 where I proposed adding the actual Apple Fn key. Unfortunately there is something of a drawback to it (you have to use an Apple VID/PID, and modify the QMK internals a little) so for legal reasons it wasn't implemented officially, however there is a diff if you want to follow along.

**Comment by fauxpark** [comment]
@drashna it seems as though we could free up more space by replacing instances of `KC_FN*` with `F(*)` as they appear to end up doing the exact same thing in `action_for_key()` (and `F()` allows for a full 12 bits worth of macros vs 5 bits with the FN* keycodes):

```c
        case KC_FN0 ... KC_FN31:
            action.code = keymap_function_id_to_action(FN_INDEX(keycode));
            break;
      // ... 
       case QK_FUNCTION ... QK_FUNCTION_MAX: ;
            // Is a shortcut for function action_layer, pull last 12bits
            // This means we have 4,096 FN macros at our disposal
            action.code = keymap_function_id_to_action( (int)keycode & 0xFFF );
            break;
```

File: tmk_core/common/report.h [hunk 1]
```diff
@@ -194,6 +197,8 @@ typedef struct {
     (key == KC_WWW_REFRESH      ?  AC_REFRESH : \
     (key == KC_BRIGHTNESS_UP    ?  BRIGHTNESSUP : \
     (key == KC_BRIGHTNESS_DOWN  ?  BRIGHTNESSDOWN : \
+    (key == KC_LAUNCHPAD        ?  LAUNCHPAD : \
```
Comment (drashna) [inline tmk_core/common/report.h]: If the order matters, then launchpad should be last.

File: tmk_core/common/report.h [hunk 2]
```diff
@@ -194,6 +197,8 @@ typedef struct {
...
+    (key == KC_MISSION_CONTROL  ?  MISSION_CONTROL : \
     (key == KC_WWW_FAVORITES    ?  AC_BOOKMARKS : 0)))))))))))))))))))))))
```
Comment (drashna) [inline tmk_core/common/report.h]: You would need to add two more close parenthesis here.

**Comment by lbibass** [comment]
okay, that makes a little more sense. I assume that there is no way to get rid of fn0? Is it necessary in QMK?

**Comment by fauxpark** [comment]
Removing the FN keycodes right now will break many keymaps which still use them. They must all be replaced with `F()` as I mentioned above.

Also, the various HID descriptors currently only declare consumer usages up to 0x029C, which probably explains why you are still not getting the correct behaviour.

**Comment by grota** [comment]
Are we really limited to 255 keycodes? Can't we "expand" the cardinality or maybe use another "space"?

**Comment by fauxpark** [comment]
Keycodes in QMK are actually 16 bits wide, but even then there are few gaps because a lot of QMK features take keycodes from the basic set as a parameter in the low byte.

Have a look at quantum_keycodes.h ;)

**Comment by lbibass** [comment]
So are you saying to add those keycodes to quantum_keycodes.h @fauxpark? That does seem pretty interesting, but I am not sure as to how I would do that.

**Comment by grota** [comment]
wouldn't it be better to move all our Consumer Page keycodes into `quantum_keycodes.h`?

My rationale is as follows:

- I don't think they need to be Shift-ed, Alt-ed and so on.
- There are potentially *a lot* of them in the spec, we might want to add more than 30 and we will run out of space again.
- The basic set is already crowded and we would give it some space to breathe.

**Comment by grota** [comment]
yep. Not right now because I have to study some things for work, but I'll look into it "soonish" :smile:

**Comment by grota** [comment]
@fauxpark just out of curiosity, I did not understand this segment here: `1000 - 10FF: Basic keycodes with right flag set but no modifiers` what does it represent again?

**Comment by fauxpark** [comment]
It's just a side effect of how the modifiers are represented in keycodes, which unfortunately takes up those 256 keycode slots. They act exactly like unmodified keycodes because they have the "right modifier" flag set (`0x1000`) but no modifiers (the set bits in `0x0F00`). For all intents and purposes they are useless.

**Comment by drashna** [comment]
There are a number of merge conflicts here. 

Alternatively, is it okay if we close the PR?

**Pull request closed without merging**